WIP: support customisations

chriadam opened the pull request.

chriadam commented on Global.qml:
```diff
@@ -115,5 +115,101 @@ QtObject {
+		},
+		{
+			"type": 2,
+			"productId": "0x106",
```
Scylla battery available in mock mode

mr-manuel commented on examples/DeviceList/README.txt:
```diff
@@ -0,0 +1,7 @@
+Run:
+`python3 ../../tools/customisationcompiler.py \
```
I would remove the backticks, so it's less confusing, also in the other readme.

If I execute this, I also get an error:

```bash
$ `python3 ../../tools/customisationcompiler.py \
    --name DeviceListExample \
    --minrequiredversion v1.2.7 \
    --devicelist 0x106 DeviceListExample_CellVoltages.qml devicelistexample_cellvoltages_title_cell_voltages \
    --devicelist 0x106 DeviceListExample_CellTemperatures.qml 'Cell Temperatures' \
    --devicelist 0x106 DeviceListExample_ComponentError.qml 'Invalid Customisation'`
Traceback (most recent call last):
  File "/mnt/c/Users/Manuel/Nextcloud/01_Privat/09_GitHub/venus-os_gui-v2__victronenergy/examples/DeviceList/../../tools/customisationcompiler.py", line 142, in <module>
    run_lupdate(qmlFiles, tsFiles, args.name)
  File "/mnt/c/Users/Manuel/Nextcloud/01_Privat/09_GitHub/venus-os_gui-v2__victronenergy/examples/DeviceList/../../tools/customisationcompiler.py", line 27, in run_lupdate
    result = subprocess.run(cmd, capture_output=True, text=True, check=True)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/subprocess.py", line 548, in run
    with Popen(*popenargs, **kwargs) as process:
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/subprocess.py", line 1026, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "/usr/lib/python3.12/subprocess.py", line 1955, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
FileNotFoundError: [Errno 2] No such file or directory: 'lupdate'
---: command not found
```

You have to install (`./scripts/build-gx-install-requirements.sh`) and source the SDK first.

mr-manuel commented:
Once I have generated the JSON file, how do I apply it, that it's visible on the local GX screen and in the WebAssembly build?

> It also includes models of customisations, which are loaded at runtime from a given settings path:
> /Gui2/Customisations
> and the names/identifiers of enabled customisations:
> /Gui2/EnabledCustomisations

I assume you mean the dbus settings path and not some settings path on the filesystem, right? Would be nice, if you can include all needed steps in the EXAMPLE.txt files. Currently I don't know how to make that work.

If there are multiple customisations from different apps, how does that work? It that is a dbus-path, then any app will overwrite the `/Gui2/Customisations` and `/Gui2/EnabledCustomisations`?

For what do I need the `QRC`, `RCC`, `QM` and `TS` files and where do I have to put them?

chriadam commented:
In practise, Matthijs mentioned that the user will put the compiled .json file into some directory on the device's filesystem, and then some service (maybe venus-platform) will pick those .json files up and inject them (in an array) to the dbus settings path mentioned.

For testing / debugging, you can just inject the content of your own .json file directly into the Global.qml, e.g. by adding a new string property for your own json content (e.g. `property string debug_myCustomisation: <json content>`) and then including that one also in the `debugJson` array property (that one "simulates" the sort of value we would normally expect to find in /Gui2/Customisations setting).

chriadam commented on src/customisations.cpp:
```diff
@@ -0,0 +1,699 @@
+		const QVector<CustomisationIntegration> integrations = c.integrations();
+		for (const CustomisationIntegration i : integrations) {
+			if ((i.type() == m_type || m_type == Customisations::InvalidIntegrationType)
+					&& (i.productId() == m_productId || m_productId.isEmpty())) {
```
should use a case-insensitive comparison here

mr-manuel commented:
Thanks, now I'm able to add the page with the functionality of this PR and they look like the pages I added manually before. The device ID now also can be entered lower-/uppercase.

It could be handy to concatenate translation strings, use common words for the menu entry or add custom code for it, like:

```
--devicelist 0xBA77 PageBatteryDbusSerialbattery.qml 'dbus-serialbattery - General'
```

Since `General` is already translated in the main Venus OS, it would be nice, if it's possible to do something like:

```
--devicelist 0xBA77 PageBatteryDbusSerialbattery.qml '"dbus-serialbattery - " + qsTrId("settings_general")'
```

But that could easily break the page where the menu entry is shown, if there is an error. You think it's better to leave it as it is?

mpvader commented:
> Not so critical, but if I have the device page opened and I change the language, all items in the menu change language except for the custom items. Same if I have a custom page opened. I have then to go back one page and open the page again, then the language is correctly displayed.

hi, in case improving things like this adds code / complexity, then I prefer to leave it as is. Its no problem for users to (even) have to reboot in case they change the language.

chriadam commented:
> Not so critical, but if I have the device page opened and I change the language, all items in the menu change language except for the custom items. Same if I have a custom page opened. I have then to go back one page and open the page again, then the language is correctly displayed.

It should work already, so long as the title you pass in the call to the compiler is a translation catalogue id.
e.g. see the DeviceListExample entry for the CellVoltages page, it looks like `--devicelist 0x106 DeviceListExample_CellVoltages.qml devicelistexample_cellvoltages_title_cell_voltages` where `devicelistexample_cellvoltages_title_cell_voltages` is a translation id in the catalogue (but there is no finished translation in the _fr.ts for it, so currently it won't actually retranslate; but if an entry existed it would).

mr-manuel commented:
I made some test with my driver mods and the translations are working, as long as I go one page back and open the page again, but it‘s nothing trivial. Maybe we should leave it as it is.

chriadam commented:
Ready for review.

blammit commented on components/DevicePage.qml:
```diff
@@ -80,6 +80,29 @@ Page {
+						required property string url
+						required property string title
+						text: qsTrId(title)
+						indicatorColor: Customisations.customisation(customisationName).color
```
Is there a particular reason for not passing in the indicator colour as a required property?

chriadam replied:
good point, will do

blammit commented on components/DevicePage.qml:
```diff
@@ -80,6 +80,29 @@ Page {
 			}
+
+			SettingsColumn {
+				id: customisationsColumn
```
id is unused

blammit commented on components/CustomDevicePageEntry.qml:
```diff
@@ -0,0 +1,6 @@
+import QtQuick
```
Oh right, it's a page linked to an entry in a DevicePage. 

What about linking it to the "customisations" terminology and calling it CustomisationDevicePage?

chriadam replied:
Well, it's not a device page.  It's a custom entry in an otherwise normal device page.

blammit replied:
But this is defining a page and the contents inside it; it just happens that the page will be linked to as a list entry from another page. This is not defining how the entry looks within the device page list view.

chriadam commented on components/listitems/core/SettingsListNavigation.qml:
```diff
@@ -14,9 +14,10 @@ ListNavigation {
 	property string iconSource: ""
 	property alias text: primary.text
 	property alias secondaryText: secondary.text
+	property var pageProperties: ({"title": Qt.binding(function() { return root.text }) })
```
This doesn't change how title is treated, it's just to ensure we don't regress (i.e. the code previously pushed these page properties already) while allowing it to be extended (as we need to pass different properties in the customisation case, at least).

blammit replied:
Ah yes, I see

chriadam commented on pages/settings/PageSettingsIntegrations.qml:
```diff
@@ -282,6 +282,45 @@ Page {
+						required property string name
+						required property color color
+						required property var integrations
+						readonly property var customisationSettingsPageIntegration: {
```
No, I don't want to do that, otherwise we'd potentially have to hoist ALL things associated with different integrations up to the top level.  The current API basically says: A Customisation can provide one or more integrations; each integration can provide a bunch of custom stuff in an integration-specific way / with an integration-specific API.

chriadam commented on src/customisations.cpp:
```diff
@@ -0,0 +1,723 @@
+	Language *languageSingleton = Language::create();
+	connect(languageSingleton, &Language::currentLanguageChanged,
+		this, [this, languageSingleton] {
+			for (const Customisation &c : qAsConst(m_customisations)) {
```
true, I will fix the ones I have added here, in this PR.

chriadam commented on src/customisations.cpp:
```diff
@@ -0,0 +1,723 @@
+	m_customisationTranslators.clear();
+}
+
+bool Customisations::installCustomisationTranslatorForLanguage(const QString &customisationName, QLocale::Language language)
```
I don't have a strong feeling either way.  On the one hand, I like that all customisation code is entirely separate from the rest of the codebase, especially since we do special stuff with translations from customisations (e.g. loading and unloading them dynamically on demand).  But on the other hand, you're right that there is some code duplication (but note that there are also some slight differences in handling, which are deliberate).  We could unify them, yes, but I'm not entirely convinced that it would be better to do so, in order to save a small amount of duplication.

blammit replied:
Okay, seems fine to leave it separate.

chriadam commented on src/customisations.h:
```diff
@@ -0,0 +1,304 @@
+
+There are currently 5 supported types of integrations:
+  1) a customisation settings page
+     - causes the customisation entry in Settings/Integrations/Customisations
```
Will fix.

chriadam commented on src/customisations.h:
```diff
@@ -0,0 +1,304 @@
+	QVector<CustomisationIntegration> m_integrations;
+};
+
+class CustomisationsModel : public QAbstractListModel, public QQmlParserStatus
```
Yep, will change

chriadam replied:
renamed to GuiPluginModel

chriadam commented on tools/customisationcompiler.py:
```diff
@@ -0,0 +1,205 @@
+
+def b64_encode_rcc(name):
+    rccFile = "" + name + ".rcc"
+    base64Str = ""
```
As in, you don't like camel-cased variable names in Python?  Do you think I should use rccfile instead of rccFile (or perhaps rcc_file, and then presumably also base64_str)?

chriadam replied:
good catch, thanks!

blammit commented on src/customisations.cpp:
```diff
@@ -0,0 +1,723 @@
+
+			if (integrationType == Customisations::InvalidIntegrationType
+					|| integrationType > Customisations::QuickAccessPaneCard
+					|| (integrationType == 2
```
Maybe an enum just inside this class then? It seems like it would be more readable, but no strong opinion either way

chriadam replied:
updated to use enum values

chriadam commented:
Updated with a variety of small changes, as per review.  Also changed the directories from which the plugins are loaded, as per discussion during meetings yesterday.  I think this one is ready for final review then merge (with this feature being Tech Preview / unsupported but available in Beta).

DanielMcInnes commented on src/guiplugins.cpp:
```diff
@@ -0,0 +1,822 @@
+		if (f.open(QIODevice::ReadOnly)) {
+			const QString plugin = QString::fromUtf8(f.readAll());
+			if (!plugin.isEmpty()) {
+				plugins.append(plugin);
```
Do we need to worry about removing plugins if they get deleted from the filesystem?

chriadam replied:
Yes.  This should work already.  We unload all plugin data when the filesystem watcher gets triggered, and then reload everything.

DanielMcInnes commented on src/guiplugins.cpp:
```diff
@@ -0,0 +1,822 @@
+
+int GuiPluginModel::rowCount(const QModelIndex &) const
+{
+	return static_cast<int>(m_plugins.count());
```
return parent.isValid() ? 0 : m_plugins.count();

DanielMcInnes commented on src/guiplugins.cpp:
```diff
@@ -0,0 +1,822 @@
+
+int GuiPluginIntegrationModel::rowCount(const QModelIndex &) const
+{
+	return static_cast<int>(m_integrations.count());
```
return static_cast<int>(m_integrations.count());

chriadam replied:
I don't understand this one, did github eat angle brackets?

DanielMcInnes reviewed (approved):
lgtm; nits

chriadam commented on src/guiplugins.cpp:
```diff
@@ -0,0 +1,822 @@
+
+		GuiPlugin p;
+		p.m_name = pluginName;
+		p.m_color = determineColor(pluginName);
```
I see - we check plugins() in determineColor() (i.e. m_plugins) but we use temporary variables here - so if we are dealing with multiple plugins in this loop, determineColor() won't be aware of them.  Thank you, I will fix.
